rbd/rbd_sparsify.go

```diff
@@ -0,0 +1,82 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rbd
+
+/*
+#cgo LDFLAGS: -lrbd
+#include <errno.h>
+#include <stdlib.h>
+#include <rbd/librbd.h>
+#include <rados/librados.h>
+
+extern int sparsifyCallback(uint64_t, uint64_t, uintptr_t);
+
+// inline wrapper to cast uintptr_t to void*
+static inline int wrap_rbd_sparsify_with_progress(rbd_image_t image,
+		size_t sparse_size, uintptr_t arg) {
+	return rbd_sparsify_with_progress(
+		image, sparse_size, (librbd_progress_fn_t)sparsifyCallback, (void*)arg);
+};
+*/
+import "C"
+
+import (
+	"github.com/ceph/go-ceph/internal/callbacks"
+)
+
+// SparsifyCallback defines the function signature needed for the
+// SparsifyWithProgress callback.
+//
+// This callback will be called by SparsifyWithProgress when it
+// wishes to report progress on sparse.
+type SparsifyCallback func(uint64, uint64, interface{}) int
+
+var sparsifyCallbacks = callbacks.New()
+
+type sparsifyCallbackCtx struct {
+	callback SparsifyCallback
+	data     interface{}
+}
+
+// SparsifyWithProgress makes an image sparse by deallocating runs of zeros.
+// The sparseSize value will be used to find runs of zeros and must be
+// a power of two no less than 4096 and no larger than the image size.
+// The given progress callback will be called to report on the progress
+// of sparse.
```

Comment:
IIRC, returning a non-zero value from the callback will also abort the function. It may be good to (a) verify that I'm correct and if so (b) mention it in the doc comment.

Reply:
Yeah, it does return a `EPERM` error.
Updated the doc.